New transform: broadcast_along_unused_hw_axes

kaushikcfd opened the pull request:
Step 1 of reviving #120 to get us a softer version of purging boostability.

kaushikcfd commented:
All the tests pass now, ready for review.

inducer commented on loopy/transform/iname.py:
```diff
@@ -1789,4 +1791,108 @@ def add_inames_to_insn(knl, inames, insn_match):
+    # if multiple inames are tagged with g.i
+    group_axes_to_inames = []
+
+    # {{{ fil axes_to_inames
```
```suggestion
    # {{{ fill axes_to_inames
```

inducer commented on loopy/transform/iname.py:
```diff
@@ -1789,4 +1791,108 @@ def add_inames_to_insn(knl, inames, insn_match):
+        raise LoopyError("Kernels containing l.auto tags are invalid"
+                " arguments.")
+
+    # local_axes_to_inames: ith entry contains the iname tagged with l.i or None
+    # if multiple inames are tagged with l.i
+    local_axes_to_inames = []
+    # group_axes_to_inames: ith entry contains the iname tagged with g.i or None
+    # if multiple inames are tagged with g.i
+    group_axes_to_inames = []
```
Move into comment block

kaushikcfd replied:
0216d54

inducer commented on loopy/transform/iname.py:
```diff
@@ -1789,4 +1791,108 @@ def add_inames_to_insn(knl, inames, insn_match):
+
+            for axis in missing_local_axes:
+                iname = local_axes_to_inames[axis]
+                if iname:
```
```suggestion
                if iname is not None:
```
to more explicitly reflect the above.

kaushikcfd replied:
0216d54

inducer commented on loopy/transform/iname.py:
```diff
@@ -1789,4 +1791,108 @@ def add_inames_to_insn(knl, inames, insn_match):
+
+            for axis in missing_group_axes:
+                iname = group_axes_to_inames[axis]
+                if iname:
```
```suggestion
                if iname is not None:
```

kaushikcfd commented on loopy/transform/iname.py:
```diff
@@ -1789,4 +1791,108 @@ def add_inames_to_insn(knl, inames, insn_match):
 # }}}
 
 
+def broadcast_along_unused_hw_axes(kernel, within=None):
```
"redundantly" seems inaccurate to me. For ex. if a instruction with unused hw axes is writing to a variable in the private address space then executing along the unused axes isn't redundant.

kaushikcfd replied:
How about just "add_unused_hw_axes"?

inducer replied:
Yeah, not bad. How about `add_inames_for_unused_hw_axes`?

kaushikcfd replied:
Yep, we have a winner :). See 0216d54

kaushikcfd commented:
Apart from addressing review comments, added one more commit ccb8e8a to address [pylint CI failure](https://github.com/inducer/loopy/pull/123/checks?check_run_id=874608014).

inducer commented on loopy/transform/iname.py:
```diff
@@ -1789,4 +1791,110 @@ def add_inames_to_insn(knl, inames, insn_match):
 
+def add_inames_for_unused_hw_axes(kernel, within=None):
+    """
+    Returns a kernel with added inames corresponding to the unused hw axes for
+    each instruction.
```
```suggestion
    Returns a kernel with inames added to each instruction
    corresponding to any hardware-parallel iname tags
    (:class:`loopy.kernel.data.GroupIndexTag`,
    :class:`loopy.kernel.data.LocalIndexTag`) unused
    in the instruction but used elsewhere in the kernel.
```

kaushikcfd commented:
> I'll apply the suggestion and merge when CI passes.

Thank you!
